j5/components/gpio_pin.py

```diff
@@ -0,0 +1,194 @@
+"""Classes for GPIO Pins."""
+
+from abc import abstractmethod
+from enum import IntEnum
+from typing import List, Optional, Type
+
+from j5.boards import Board
+from j5.components import Component, Interface, NotSupportedByHardwareError
+
+
+class BadGPIOPinModeError(Exception):
+    """The pin is not in the correct mode."""
+
+    pass
+
+
+class GPIOPinMode(IntEnum):
+    """Hardware modes that a GPIO pin can be set to."""
+
+    DIGITAL_INPUT = 0  #: The digital state of the pin can be read
+    DIGITAL_INPUT_PULLUP = 1  #: Same as DIGITAL_INPUT but internal pull-up is enabled
+    DIGITAL_INPUT_PULLDOWN = 2  #: Same as DIGITAL_INPUT but internal pull-down is enabled
+    DIGITAL_OUTPUT = 3  #: The digital state of the pin can be set.
+
+    ANALOGUE_INPUT = 4  #: The analogue voltage of the pin can be read.
+    ANALOGUE_OUTPUT = 5  #: The analogue voltage of the pin can be set using a DAC.
+
+    PWM_OUTPUT = 6  #: A PWM output signal can be created on the pin.
+
+
+class GPIOPinInterface(Interface):
+    """An interface containing the methods required for a GPIO Pin."""
+
+    @abstractmethod
+    def set_gpio_pin_mode(self,
+                          board: Board,
+                          identifier: int,
+                          pin_mode: GPIOPinMode,
+                          ) -> None:
+        """Set the hardware mode of a GPIO pin."""
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def get_gpio_pin_mode(self, board: Board, identifier: int) -> GPIOPinMode:
+        """Get the hardware mode of a GPIO pin."""
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def write_gpio_pin_digital_state(self,
+                                     board: Board,
+                                     identifier: int,
+                                     state: bool,
+                                     ) -> None:
+        """Write to the digital state of a GPIO pin."""
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def get_gpio_pin_digital_state(self, board: Board, identifier: int) -> bool:
+        """Get the last written state of the GPIO pin."""
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def read_gpio_pin_digital_state(self, board: Board, identifier: int) -> bool:
+        """Read the digital state of the GPIO pin."""
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def read_gpio_pin_analogue_value(self, board: Board, identifier: int) -> float:
+        """Read the scaled analogue value of the GPIO pin."""
```

Comment:
What is a "scaled analogue value" in this context?

Reply:
A float between `0` and `1` that is scaled such that `scaled_value * max_voltage = current_voltage`

Reply:
I agree that more detailed docs are required for competitor purposes.